File: Darts/src/code/db/DartsMatchEntity.java
package code.db;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.Collections;
import java.util.HashMap;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import code.utils.DatabaseUtil;
import object.HandyArrayList;
import util.DateUtil;
import util.Debug;
import util.XmlUtil;

/**
 * Simple entity to join multiple 'games' together into a 'match'.
 * Table has to be called 'DartsMatch' because 'Match' is a derby keyword!
 */
public class DartsMatchEntity extends AbstractDartsEntity<DartsMatchEntity>
{
	public static final int MODE_FIRST_TO = 0;
	public static final int MODE_POINTS = 1;
	
	private int games = -1;
	private int mode = -1;
	private Timestamp dtFinish = DateUtil.END_OF_TIME;
	private String matchParams = "";
	
	//Non-db things
	private int currentOrdinal = 0;
	private String gameParams = "";
	private int gameType = -1;
	private HandyArrayList<PlayerEntity> players = new HandyArrayList<>();
	private HashMap<Integer, Integer> hmPositionToPoints = null;
	
	@Override
	public String getTableName()
	{
		return "DartsMatch";
	}

	@Override
	public String getCreateTableSqlSpecific()
	{
		return "Games INT NOT NULL, Mode INT NOT NULL, DtFinish TIMESTAMP NOT NULL, MatchParams VARCHAR(255) NOT NULL";
	}

	@Override
	public void populateFromResultSet(DartsMatchEntity entity, ResultSet rs)
			throws SQLException
	{
		entity.setGames(rs.getInt("Games"));
		entity.setMode(rs.getInt("Mode"));
		entity.setDtFinish(rs.getTimestamp("DtFinish"));
		entity.setMatchParams(rs.getString("MatchParams"));
	}

	@Override
	public String writeValuesToStatement(PreparedStatement statement, int i, String statementStr) throws SQLException
	{
		statementStr = writeInt(statement, i++, games, statementStr);
		statementStr = writeInt(statement, i++, mode, statementStr);
		statementStr = writeTimestamp(statement, i++, dtFinish, statementStr);
		statementStr = writeString(statement, i++, matchParams, statementStr);
		return statementStr;
	}
	
	@Override
	public long getGameId()
	{
		return -1;
	}
	
	/**
	 * Helpers
	 */
	public int getScoreForFinishingPosition(int position)
	{
		if (mode == MODE_FIRST_TO)
		{
			return position == 1? 1:0;
		}
		
		if (mode == MODE_POINTS)
		{
			if (position == -1)
			{
				return 0;
			}
			
			return getHmPositionToPoints().get(position);
		}
		
		Debug.stackTrace("Unimplemented for match mode [" + mode + "]");
		return -1;
	}
	public boolean isComplete()
	{
		if (mode == MODE_FIRST_TO)
		{
			return isFirstToMatchComplete();
		}
		else if (mode == MODE_POINTS)
		{
			return isPointsMatchComplete();
		}
		
		Debug.stackTrace("Unimplemented for match mode [" + mode + "]");
		return false;
	}
	private boolean isFirstToMatchComplete()
	{
		StringBuilder sb = new StringBuilder();
		sb.append(" SELECT COUNT(1) AS WinCount");
		sb.append(" FROM Participant p, Game g");
		sb.append(" WHERE g.DartsMatchId = ");
		sb.append(getRowId());
		sb.append(" AND p.GameId = g.RowId");
		sb.append(" AND p.FinishingPosition = 1");
		sb.append(" GROUP BY p.PlayerId");
		sb.append(" ORDER BY COUNT(1) DESC");
		sb.append(" FETCH FIRST 1 ROWS ONLY");
		
		try (ResultSet rs = DatabaseUtil.executeQuery(sb))
		{
			if (rs.next())
			{
				int count = rs.getInt("WinCount");
				return count == games;
			}
		}
		catch (SQLException sqle)
		{
			Debug.logSqlException(sb.toString(), sqle);
		}
		
		return false;
	}
	private boolean isPointsMatchComplete()
	{
		StringBuilder sb = new StringBuilder();
		sb.append(" SELECT COUNT(1)");
		sb.append(" FROM Game");
		sb.append(" WHERE DartsMatchId = ");
		sb.append(getRowId());
		sb.append(" AND DtFinish < ");
		sb.append(DateUtil.getEndOfTimeSqlString());
		
		long count = DatabaseUtil.executeQueryAggregate(sb);
		return count == games;
	}
	public int incrementAndGetCurrentOrdinal()
	{
		return currentOrdinal++;
	}
	
	public int getPlayerCount()
	{
		return players.size();
	}
	
	public void shufflePlayers()
	{
		if (players.size() == 2)
		{
			players = players.reverse();
		}
		else
		{
			
			Collections.shuffle(players);
		}
	}
	
	public String getMatchDesc()
	{
		return "Match #" + getRowId() 
			+ " (" + getMatchTypeDesc() + " - " 
			+ GameEntity.getTypeDesc(gameType, gameParams) + ", " 
			+ getPlayerCount() + " players)";
	}
	private String getMatchTypeDesc()
	{
		if (mode == MODE_FIRST_TO)
		{
			return "First to " + games;
		}
		
		if (mode == MODE_POINTS)
		{
			return "Points based (" + games + " games)";
		}
		
		return "";
		
	}
	
	private HashMap<Integer, Integer> getHmPositionToPoints()
	{
		if (hmPositionToPoints == null)
		{
			hmPositionToPoints = new HashMap<>();
			
			Document doc = XmlUtil.getDocumentFromXmlString(matchParams);
			Element root = doc.getDocumentElement();
			
			hmPositionToPoints.put(1, XmlUtil.getAttributeInt(root, "First"));
			hmPositionToPoints.put(2, XmlUtil.getAttributeInt(root, "Second"));
			hmPositionToPoints.put(3, XmlUtil.getAttributeInt(root, "Third"));
			hmPositionToPoints.put(4, XmlUtil.getAttributeInt(root, "Fourth"));
		}
		
		return hmPositionToPoints;
	}
	
	public void cacheMetadataFromGame(GameEntity game)
	{
		this.gameType = game.getGameType();
		this.gameParams = game.getGameParams();
		this.players = game.retrievePlayersVector();
		
		//Should've been setting this too...
		this.currentOrdinal = game.getMatchOrdinal();
	}
	
	/**
	 * Factory methods
	 */
	public static DartsMatchEntity factoryFirstTo(int games)
	{
		return factoryAndSave(games, MODE_FIRST_TO, "");
	}
	public static DartsMatchEntity factoryPoints(int games, String pointsXml)
	{
		return factoryAndSave(games, MODE_POINTS, pointsXml);
	}
	private static DartsMatchEntity factoryAndSave(int games, int mode, String matchParams)
	{
		DartsMatchEntity matchEntity = new DartsMatchEntity();
		matchEntity.assignRowId();
		matchEntity.setMode(mode);
		matchEntity.setGames(games);
		matchEntity.setMatchParams(matchParams);
		matchEntity.saveToDatabase();
		return matchEntity;
	}

	/**
	 * Gets / Sets
	 */
	public int getGames()
	{
		return games;
	}
	public void setGames(int games)
	{
		this.games = games;
	}
	public int getMode()
	{
		return mode;
	}
	public void setMode(int mode)
	{
		this.mode = mode;
	}
	public Timestamp getDtFinish()
	{
		return dtFinish;
	}
	public void setDtFinish(Timestamp dtFinish)
	{
		this.dtFinish = dtFinish;
	}
	public String getMatchParams()
	{
		return matchParams;
	}
	public void setMatchParams(String matchParams)
	{
		this.matchParams = matchParams;
	}

	/**
	 * Non-db gets / sets
	 */
	public String getGameParams()
	{
		return gameParams;
	}
	public void setGameParams(String gameParams)
	{
		this.gameParams = gameParams;
	}
	public int getGameType()
	{
		return gameType;
	}
	public void setGameType(int gameType)
	{
		this.gameType = gameType;
	}
	public HandyArrayList<PlayerEntity> getPlayers()
	{
		return players;
	}
	public void setPlayers(HandyArrayList<PlayerEntity> players)
	{
		this.players = players;
	}
}

File: Darts/src/burlton/dartzee/code/db/DartsMatchEntity.kt
package burlton.dartzee.code.db

import burlton.core.code.util.*
import burlton.dartzee.code.utils.DatabaseUtil
import burlton.desktopcore.code.util.DateStatics
import burlton.desktopcore.code.util.getEndOfTimeSqlString

/**
 * Simple entity to join multiple 'games' together into a 'match'.
 * Table has to be called 'DartsMatch' because 'Match' is a derby keyword!
 */
class DartsMatchEntity : AbstractEntity<DartsMatchEntity>()
{
    /**
     * DB Fields
     */
    var localId = -1L
    var games = -1
    var mode = -1
    var dtFinish = DateStatics.END_OF_TIME
    var matchParams = ""

    /**
     * Non-db gets / sets
     */
    var gameParams = ""
    var gameType = -1
    var players = mutableListOf<PlayerEntity>()

    private var currentOrdinal = 0
    private var hmPositionToPoints: MutableMap<Int, Int>? = null

    override fun getTableName() = "DartsMatch"

    override fun getCreateTableSqlSpecific(): String
    {
        return "LocalId INT UNIQUE NOT NULL, Games INT NOT NULL, Mode INT NOT NULL, DtFinish TIMESTAMP NOT NULL, MatchParams VARCHAR(255) NOT NULL"
    }

    override fun assignRowId(): String
    {
        localId = LocalIdGenerator.generateLocalId(getTableName())
        return super.assignRowId()
    }

    /**
     * Helpers
     */
    fun isComplete(): Boolean
    {
        return when(mode)
        {
            MODE_FIRST_TO -> getIsFirstToMatchComplete()
            MODE_POINTS -> getIsPointsMatchComplete()
            else -> {
                Debug.stackTrace("Unimplemented for match mode [$mode]")
                false
            }
        }
    }

    private fun getIsFirstToMatchComplete(): Boolean
    {
        val sb = StringBuilder()
        sb.append(" SELECT COUNT(1) AS WinCount")
        sb.append(" FROM Participant p, Game g")
        sb.append(" WHERE g.DartsMatchId = '$rowId'")
        sb.append(" AND p.GameId = g.RowId")
        sb.append(" AND p.FinishingPosition = 1")
        sb.append(" GROUP BY p.PlayerId")
        sb.append(" ORDER BY COUNT(1) DESC")
        sb.append(" FETCH FIRST 1 ROWS ONLY")

        DatabaseUtil.executeQuery(sb).use { rs ->
            if (rs.next())
            {
                val count = rs.getInt("WinCount")
                return count == games
            }
        }

        return false
    }
    private fun getIsPointsMatchComplete(): Boolean
    {
        val sb = StringBuilder()
        sb.append(" SELECT COUNT(1)")
        sb.append(" FROM Game")
        sb.append(" WHERE DartsMatchId = '$rowId'")
        sb.append(" AND DtFinish < ")
        sb.append(getEndOfTimeSqlString())

        val count = DatabaseUtil.executeQueryAggregate(sb)
        return count == games
    }

    fun getPlayerCount(): Int = players.size

    fun getMatchDesc(): String
    {
        return "Match #$localId (${getMatchTypeDesc()} - ${GameEntity.getTypeDesc(gameType, gameParams)}, ${getPlayerCount()} players)"
    }

    private fun getMatchTypeDesc(): String
    {
        return when(mode)
        {
            MODE_FIRST_TO -> "First to $games"
            MODE_POINTS -> "Points based ($games games)"
            else -> ""
        }
    }

    fun getScoreForFinishingPosition(position: Int): Int
    {
        return when(mode)
        {
            MODE_FIRST_TO -> if (position == 1) 1 else 0
            MODE_POINTS -> if (position == -1) 0 else getHmPositionToPoints()[position]!!
            else -> {
                Debug.stackTrace("Unexpected mode [$mode]")
                -1
            }
        }
    }
    private fun getHmPositionToPoints(): MutableMap<Int, Int>
    {
        if (hmPositionToPoints == null)
        {
            hmPositionToPoints = mutableMapOf()

            val doc = matchParams.toXmlDoc()
            val root = doc!!.documentElement

            hmPositionToPoints!![1] = root.getAttributeInt("First")
            hmPositionToPoints!![2] = root.getAttributeInt("Second")
            hmPositionToPoints!![3] = root.getAttributeInt("Third")
            hmPositionToPoints!![4] = root.getAttributeInt("Fourth")
        }

        return hmPositionToPoints!!
    }

    fun incrementAndGetCurrentOrdinal() = ++currentOrdinal

    fun shufflePlayers()
    {
        if (players.size == 2)
        {
            players.reverse()
        }
        else
        {
            players.shuffle()
        }
    }

    fun cacheMetadataFromGame(lastGame: GameEntity)
    {
        this.gameType = lastGame.gameType
        this.gameParams = lastGame.gameParams
        this.players = lastGame.retrievePlayersVector()

        //Should've been setting this too...
        this.currentOrdinal = lastGame.matchOrdinal
    }

    companion object
    {
        const val MODE_FIRST_TO = 0
        const val MODE_POINTS = 1

        fun constructPointsXml(first: Int, second: Int, third: Int, fourth: Int): String
        {
            val doc = XmlUtil.factoryNewDocument()
            val rootElement = doc.createRootElement("MatchParams")
            rootElement.setAttribute("First", "$first")
            rootElement.setAttribute("Second", "$second")
            rootElement.setAttribute("Third", "$third")
            rootElement.setAttribute("Fourth", "$fourth")

            return doc.toXmlString()
        }


        /**
         * Factory methods
         */
        fun factoryFirstTo(games: Int): DartsMatchEntity
        {
            return factoryAndSave(games, MODE_FIRST_TO, "")
        }

        fun factoryPoints(games: Int, pointsXml: String): DartsMatchEntity
        {
            return factoryAndSave(games, MODE_POINTS, pointsXml)
        }

        private fun factoryAndSave(games: Int, mode: Int, matchParams: String): DartsMatchEntity
        {
            val matchEntity = DartsMatchEntity()
            matchEntity.assignRowId()
            matchEntity.mode = mode
            matchEntity.games = games
            matchEntity.matchParams = matchParams
            matchEntity.saveToDatabase()
            return matchEntity
        }
    }
}
